### ISSUE TITLE
Very strange issue with generators, and inferred types

### ISSUE THREAD

**Issue by kkirby** [microsoft/TypeScript#39812]
**TypeScript Version:**  3.9.2 & 4.0.0-beta & v4.1.0-dev.20201002

**Search Terms:** generators infer

**Expected behavior:** The type of `value` should not infer to `any` in this scenario

**Actual behavior:** The type of `value` is inferred to `any`.

**Related Issues:**

**Code**
```ts
interface ValueType<T = any> {
  value: T;
  type: string;
}

function* getValue<T extends string>(
  value: T
): Generator<any, ValueType<T>, any> {
  return {
    value,
    type: "string"
  };
}

function* call<T extends any>(fn: T, args?: any) {
  return {};
}

// This does NOT work

call(function*() {
  // Type of value is any
  const value = yield* getValue("hello, world");
  console.log(value.type);

    yield* call(function*(args: any) {
      yield "asdf";
    }, value.type);

});

// These work

call(function*() {
  // Type of value is correct
  const value = yield* getValue("hello, world");
  console.log(value.type);

  yield* call(function*(args: any) {
    yield "asdf";
  }); // I'm not passing value.type

});

// I'm not wrapping the function in call()
function* a() {
  // Type of value is correct
  const value = yield* getValue("hello, world");
  console.log(value.type);

  yield* call(function*(args: any) {
    yield "asdf";
  }, value.type);
}
```

<details><summary><b>Compiler Options</b></summary>

```json
{
  "compilerOptions": {
    "noImplicitAny": true,
    "strictNullChecks": true,
    "strictFunctionTypes": true,
    "strictPropertyInitialization": true,
    "strictBindCallApply": true,
    "noImplicitThis": true,
    "noImplicitReturns": true,
    "useDefineForClassFields": false,
    "alwaysStrict": true,
    "allowUnreachableCode": false,
    "allowUnusedLabels": false,
    "downlevelIteration": false,
    "noEmitHelpers": false,
    "noLib": false,
    "noStrictGenericChecks": false,
    "noUnusedLocals": false,
    "noUnusedParameters": false,
    "esModuleInterop": true,
    "preserveConstEnums": false,
    "removeComments": false,
    "skipLibCheck": false,
    "checkJs": false,
    "allowJs": false,
    "declaration": true,
    "experimentalDecorators": false,
    "emitDecoratorMetadata": false,
    "target": "ES2017",
    "module": "ESNext"
  }
}
```

</details>

**Playground Link:** [Provided](https://www.typescriptlang.org/play/index.html#code/JYOwLgpgTgZghgYwgAgGpwDYFcIBUCeADhADy7IC8ycI+AfMgN4CwAUMsgG6Y4BcyuANxsOYIhH4BnMFFABzYawC+bNjCwgEYYAHsQAKmRyIYdNlLkIAD0ggAJpOTTZIOXQAUIrjwkC2ASn4AcQgQaDgwHSgSGnwAGjQfAmIyOgTYhhZ2ZCgTLCgQJi8ObnM44uQxYn4AImd5Gq8lRRVWNQ0tXULgSQBlGXkzHHdSvkTzZNIMwO9zZB7xnEmSetdMr1ywfMLRiAA6KpQKY+Q6gdcaltVWdU1tPUMETAwyZGtbB2paDxgQflx0lA5JIAPz8WL+IrZTbbJjNNitNgAeiRAgAFgs7DoII4AHIAeXIAHcogBra5PDAYdy3ToPdyQrIcFECcTIHQwWY4eaOWJeBB6aRco7IfDACAYOyGYymHzuGpoiUYHQJElQSU1fyKDgCkCSHQYfbKuQjHwHcRa64cDhiiVS5CU6m0+4GdxwIGScG0RkVG3iyWnOCSOwwS4VJQJXbm4iWtrKWPI1EAQSp7M5YEVkhQavJccdNI6Lv0DKhzNRkzTwp5DqiuS0-MFYCrVFtkulJiGEHliqpKuQao1sZ1goNRp0JqjhwT2Vb9vzzq6xfdwK9+B92T9dsDwdD2uQSi1yBZAEkAOQAW2QIB0TcIQck8mF0Yg1wPikTyDPl+vTaJUDghCEI+GYoAuejzHq1DtHci7UCWTJHuWbIclWCwClAdZgA2epNrslCiv69oyp23ZKn2A52Jqe66vqhp7MaprmM+06bm2DrPAWMH0sunpfGupbWrO24hmG2QRk+U4tEAA)

**Other Details**

I've noticed that if you hover over `getValue()` and then hover over `type` it'll trick the compiler into inferring the type. I've created a video that demonstrates the issue.

[ts-issue.mp4.zip](https://github.com/microsoft/TypeScript/files/4996625/ts-issue.mp4.zip)

**Comment by danielweck** [comment]
We experience a similar issue with Redux Saga generators. The type is correctly inferred in Visual Studio Code until the current TypeScript file is saved (I just add a space, remove it, and save => this triggers a re-parse / recompile, and the type is now `any`)

**Comment by rbuckton** [comment]
This seems like it may have been addressed in TypeScript 4.7 as it no longer repros.
